crates/relayer-cosmos-mock/src/impls/basecoin.rs

```diff
@@ -0,0 +1,165 @@
+use crate::contexts::basecoin::MockBasecoin;
+use crate::traits::endpoint::BasecoinEndpoint;
+use crate::traits::handle::BasecoinHandle;
+use crate::types::error::Error;
+use crate::util::conversion::convert_tm_to_ics_merkle_proof;
+use crate::util::dummy::genesis_app_state;
+use crate::util::mutex::MutexUtil;
+
+use basecoin_app::modules::ibc::Ibc;
+use basecoin_app::modules::types::IdentifiedModule;
+use basecoin_store::context::ProvableStore;
+use basecoin_store::context::Store;
+use basecoin_store::impls::RevertibleStore;
+use basecoin_store::utils::SharedRwExt;
+
+use ibc::core::ics23_commitment::commitment::CommitmentProofBytes;
+use ibc::core::ics24_host::identifier::ChainId;
+use ibc::core::ics24_host::path::Path;
+use ibc::hosts::tendermint::IBC_QUERY_PATH;
+use ibc::Height;
+
+use tendermint::block::Height as TmHeight;
+use tendermint::time::Time;
+use tendermint::v0_37::abci::request::InitChain;
+use tendermint::v0_37::abci::request::Query;
+use tendermint::v0_37::abci::Request as AbciRequest;
+use tendermint::v0_37::abci::Response as AbciResponse;
+use tendermint_testgen::consensus::default_consensus_params;
+use tendermint_testgen::light_block::TmLightBlock;
+use tower::Service;
+
+use async_trait::async_trait;
+use std::fmt::Debug;
+
+#[async_trait]
+impl<S> BasecoinHandle for MockBasecoin<S>
+where
+    S: ProvableStore + Default + Debug,
+{
+    type Store = S;
+
+    /// Initialize the chain with the app state.
+    async fn init(&self) {
+        let app_state = serde_json::to_vec(&genesis_app_state()).expect("infallible serialization");
+
+        let request = InitChain {
+            time: Time::now(),
+            chain_id: self.chain_id.to_string(),
+            consensus_params: default_consensus_params(),
+            validators: vec![],
+            app_state_bytes: app_state.into(),
+            initial_height: TmHeight::from(1_u8),
+        };
+
+        let mut app = self.app.clone();
+
+        app.call(AbciRequest::InitChain(request))
+            .await
+            .expect("failed to initialize chain");
+
+        // Generates the genesis block
+        self.grow_blocks();
+    }
+
+    async fn begin_block(&self) {
+        let last_block = self.blocks.acquire_mutex().last().unwrap().clone();
+
+        let mut events = Vec::new();
+
+        let mut modules = self.app.modules.write_access();
+
+        for IdentifiedModule { id: _, module } in modules.iter_mut() {
+            let event = module.begin_block(&last_block.signed_header.header);
+            events.extend(event);
+        }
+    }
+
+    /// Commits the chain state.
+    async fn commit(&self) {
+        let mut modules = self.app.modules.write_access();
+
+        for IdentifiedModule { id, module } in modules.iter_mut() {
+            module
+                .store_mut()
+                .commit()
+                .expect("failed to commit to state");
+
+            let mut state = self.app.store.write_access();
```

Comment:
Small nit: could we not move `let mut state = self.app.store.write_access();` to the top of the function outside of the for loop and then not have to declare `let mut state` again in the rest of the function?

Reply:
[cd438ec](https://github.com/informalsystems/hermes/pull/3566/commits/cd438ec5775774bf71457a103726b39bba8f7edc)